src/pydvl/influence/torch/functional.py

```diff
@@ -447,6 +463,57 @@ def flat_input_empirical_loss(p: torch.Tensor):
     return hessian_mat
 
 
+def gauss_newton(
+    model: torch.nn.Module,
+    loss: Callable[[torch.Tensor, torch.Tensor], torch.Tensor],
+    data_loader: DataLoader,
+    restrict_to: Optional[Dict[str, torch.Tensor]] = None,
+):
+    r"""
+    Compute the Gauss-Newton matrix, i.e.
+
+    $$ \sum_{i=1}^N \nabla_{\theta}\ell(m(x_i; \theta), y)
+        \nabla_{\theta}\ell(m(x_i; \theta), y)^t,$$
+    for a  loss function $\ell$ and a model $m$ with model parameters $\theta$.
+
+    Args:
+        model: The PyTorch model.
+        loss: A callable that computes the loss.
+        data_loader: A PyTorch DataLoader providing batches of input data and
+            corresponding output data.
+        restrict_to: The parameters to restrict the differentiation to,
+            i.e. the corresponding sub-matrix of the Jacobian. If None, the full
+            Jacobian is used.
+
+    Returns:
+        The Gauss-Newton matrix.
+    """
+
+    per_sample_grads = create_per_sample_gradient_function(model, loss)
+
+    params = restrict_to
+    if params is None:
+        params = {k: p.detach() for k, p in model.named_parameters() if p.requires_grad}
```

Comment:
Dict comprehensions like this appear very often. Should we have a function `get_detached_params(model)` instead?

Reply:
Agree, will check

Reply:
Add a util function `get_model_parameters`, done.